diff -Nur '--exclude=.git' '--exclude=.github' '--exclude=*.pyc' ../backup/projects/global-components/components/TimeServer/include/plat/ls1046a/plat/timers.h global-components/components/TimeServer/include/plat/ls1046a/plat/timers.h
--- ../backup/projects/global-components/components/TimeServer/include/plat/ls1046a/plat/timers.h	1970-01-01 08:00:00.000000000 +0800
+++ global-components/components/TimeServer/include/plat/ls1046a/plat/timers.h	2023-03-12 14:02:06.290990499 +0800
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#pragma once
+
+#define HARDWARE_TIMER_COMPONENT                                       \
+    component HWTimer {                                                \
+        hardware;                                                      \
+        emits TimerIRQ irq;                                          \
+    }
+
+#define HARDWARE_TIMER_INTERFACES                                      \
+    consumes TimerIRQ hwtimer_irq;
+#define HARDWARE_TIMER_ATTRIBUTES
+#define HARDWARE_TIMER_COMPOSITION                                     \
+        component HWTimer hwtimer;                                    \
+        connection seL4HardwareInterrupt hwtimer_irq(from hwtimer.irq, \
+                                                      to hwtimer_irq);
+#define HARDWARE_TIMER_CONFIG                                            \
+    hwtimer.irq_irq_number = 30;                                        \
+    hwtimer.irq_irq_type = "arm";
+#define HARDWARE_TIMER_PLAT_INTERFACES
